diff -Nur network-manager-applet-0.6.5/src/applet.c network-manager-applet-0.6.5.new/src/applet.c
--- network-manager-applet-0.6.5/src/applet.c	2007-06-26 08:30:16.000000000 +0000
+++ network-manager-applet-0.6.5.new/src/applet.c	2007-06-26 08:30:21.000000000 +0000
@@ -1258,7 +1258,7 @@
 /*
  * nma_menu_network_admin_activate
  *
- * Signal function called when user clicks "Static configuration..."
+ * Signal function called when user clicks "Manual configuration..."
  *
  */
 static void nma_menu_network_admin_activate (GtkMenuItem *item, gpointer user_data)
@@ -1694,7 +1694,7 @@
 	g_return_if_fail (menu != NULL);
 	g_return_if_fail (applet != NULL);
 
-	item = GTK_MENU_ITEM (gtk_menu_item_new_with_mnemonic (_("_Static configuration...")));
+	item = GTK_MENU_ITEM (gtk_menu_item_new_with_mnemonic (_("_Manual configuration...")));
 
 	g_signal_connect (G_OBJECT (item), "activate", G_CALLBACK (nma_menu_network_admin_activate), applet);
 	gtk_menu_shell_append (GTK_MENU_SHELL (menu), GTK_WIDGET (item));
@@ -1856,7 +1856,10 @@
 
 	if (!applet->device_list)
 	{
-		nma_menu_add_text_item (menu, _("No network devices have been found"));
+		if (applet->nm_state != NM_STATE_CONNECTED)
+			nma_menu_add_text_item (menu, _("No network devices have been found"));
+
+		nma_menu_add_network_admin_item (menu, applet);
 		return;
 	}
 
